scripts: twister: Isolate statuses into a separate class

LukaszMrugala opened the pull request:
Now statuses are not just an `str` that can be easily mistyped or assigned wrong. Now they are an `Enum`.

Every object that can have a status has a separate `Status` class. It is probably an overkill, but we want a gradual change of current status handling system.

LukaszMrugala committed:
scripts: twister: Isolate statuses into a separate class

Now statuses are not just a str that can be easily mistyped
or assigned wrong. Now they are an Enum.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

nashif commented on scripts/pylib/twister/twisterlib/statuses.py:
```diff
@@ -0,0 +1,68 @@
+    ERROR = 'error'
+    FAIL = 'failed'
+    FILTER = 'filtered'
+    FLASH = 'flash_error'
```
that should not be exposed as status, needs to clean this up

LukaszMrugala replied:
It is a status currently, thus should be categorised analogously to every other status. As I assume we would like to excise it from Zephyr, we ought to do it in the next step of our status overhaul. That next step will also be easier with every use of such status highlighted.

nashif replied:
please remove it now, as it is a leftover and does not propagte as status in current reports, this should translate to simple error with a reason of flash error.  see here https://github.com/zephyrproject-rtos/zephyr/pull/61414/commits/2d9384d50aa7b265d21912296c5137712c09e2e2 and here https://github.com/zephyrproject-rtos/zephyr/pull/61414/commits/2de29afdcfc60a23d54fbd1980c1b2873b408811

LukaszMrugala replied:
Removed.

nashif commented on scripts/pylib/twister/twisterlib/statuses.py:
```diff
@@ -0,0 +1,68 @@
+    FLASH = 'flash_error'
+    PASS = 'passed'
+    SKIP = 'skipped'
+    TIMEOUT = 'timeout'
```
ditto

LukaszMrugala replied:
Removed.

LukaszMrugala committed:
scripts: twister: Remove unused statuses

TestInstanceStatus of TIMEOUT and FLASH were never really used.
They were checked for, but never assigned.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala commented on scripts/pylib/twister/twisterlib/statuses.py:
This PR was supposed to be very limited in its scope. Your suggestion, with guards against wrong assignments, seems more extensive and thus suitable for a separate Issue and PR.

golowanow replied:
> .. or assigned wrong.

just changing strings to enum doesn't prevent of wrong assignment without transition check.

Also, with setters is should be possible to convert strings to enum in one place, at the setter, and raise exception on a wrong value.

LukaszMrugala replied:
The plan for this PR was to only change strings into Enums, without any additional functionalities. The benefits are: easier use of search function and more visible status organisation. 
Your suggestions sound useful, but I'd rather first create the classes and then add those functionalities for them.

LukaszMrugala replied:
@golowanow I have attempted a revised approach, with dual setter guards against incorrect assignments.

nashif commented on scripts/pylib/twister/twisterlib/statuses.py:
```diff
@@ -0,0 +1,84 @@
+    NONE = None  # to preserve old functionality
+    FILTER = 'filtered'
+    PASS = 'passed'
+    SKIP = 'skipped'
```
Suites do not fail?

LukaszMrugala replied:
`TestInstance` can have a `FAIL` status and that status can be assigned to `TestSuite` - thus, `TestSuite` can only fail if its `TestInstance` fails.

LukaszMrugala replied:
Unified Status classes into one.

nashif commented on scripts/pylib/twister/twisterlib/statuses.py:
```diff
@@ -0,0 +1,84 @@
+
+# Possible direct assignments:
+#   * OutputStatus <- HarnessStatus
+class OutputStatus(str, Enum):
```
I do not think any of those are statuses, more of reasons?

LukaszMrugala replied:
Those explicitly declared here are mostly kinds of `ERROR`. However, any `HarnessStatus` can be assigned as `output_status`.

nashif commented on scripts/pylib/twister/twisterlib/reports.py:
```diff
@@ -51,34 +53,34 @@ def xunit_testcase(eleTestsuite, name, classname, status, ts_status, reason, dur
                 tc_type = status
-            ET.SubElement(eleTestcase, 'skipped', type=f"{tc_type}", message=f"{reason}")
-        elif status in ["failed", "blocked"]:
+            ET.SubElement(eleTestcase, ReportStatus.SKIP, type=f"{tc_type}", message=f"{reason}")
```
I would leave those sepeare from the twister status, we need to be able to translate in the report from twister status to junit/xunit but I would not manage them at the same level as the other statuses

LukaszMrugala replied:
I have excised that class from the rest of the statuses and moved it to `reports.py`. Thus we can achieve clear distinction from the rest of the Status classes while helping code readers know which values are possible.

nashif reviewed (changes requested):
what is the motivation for having many status classes for each step of the execution process, why cant all of those be abstracted with one status class?

LukaszMrugala commented:
> what is the motivation for having many status classes for each step of the execution process, why cant all of those be abstracted with one status class?

Should we add some checks for status assignment, it would be easier if we differentiate different kinds of statuses.

hakehuang commented on scripts/pylib/twister/twisterlib/statuses.py:
```diff
@@ -0,0 +1,91 @@
+    SKIP = 'skipped'
+
+
+class TestSuiteStatus(str, Enum):
```
why not just do below, the reasons are:
1. rephase the enumrate from before seems confusing.
2. give a way that we can just judge the return value to see it pass and failed
`
class TestSuiteStatus(str, Enum):
    none = None
    passed = 0
    filtered = 1
    skipped = 2
    error = 3
    failed = 4

    def __str__(self):
        return str(self.name)

`

LukaszMrugala replied:
There are possible problems with both `none` and `passed` being falsey while the others are truthy.

Technically it would be a change to results, as the old "None" statuses would now be "none" statuses (Minor, could be changed by a check in `__str__`, but still warrants a mention).

This would tie the status `str` representation strongly to the code representation, thus making further changes harder (What if someone decides that the fact that all statuses use past participle save for `error` is unacceptable and it should be changed to `errored`? We would need to change every use of the `TestSuiteStatus.error`.)

We currently have no use for checking just the return value instead of the status name. Doing so would make the code less readable (Compare: `status in [Status.SKIP, Status.FILTER]` vs `return_code in [1, 2]`)

nashif commented:
> Should we add some checks for status assignment, it would be easier if we differentiate different kinds of statuses.

would be great to see where this is going with more details, this seems to be a consideration for possible future enhacements that we might not do at all? So if we you have a proposal of how this would look like, it is worth showing the value now.

nashif reviewed (changes requested):
more comments below. Still not convinced with the multiple status classes and how they different and have custom reason strings promoted to be statuses.

LukaszMrugala commented on scripts/pylib/twister/twisterlib/statuses.py:
```diff
@@ -0,0 +1,91 @@
+
+    # Values assigned directly
+    NONE = None  # to preserve old functionality
+    BYTE = 'unexpected byte'
```
This commit was planned only to isolate all the different statuses and their assignments, deduce what statuses are currently in use and separate them into a separate class. It was a descriptive change that describes how we currently approach statuses, just in an easier-to-read way.

Status-assignment code was borne out of a desire to protect the programmers from incorrect status assignments. Note that had we had that before, catching the strange QEMU statuses that honestly should just be an `ERROR` with a proper "reason" field could have been easier.

If we merge all the statuses into one, then why is there a `STARTED` status that is assignable for e.g. `Harness`? What does it mean for `Harness` to be `BLOCK`? Without separate status classes and without a separate property assignment code in status-having classes, we invite such errors. This could be resolved with a consistent status policy, where all statuses make sense on all classes that can have a status, but that would be even more changes to the status ecosystem in one big PR. At that point, it might be easier to revive #61414 directly.

nashif replied:
see https://github.com/zephyrproject-rtos/zephyr/pull/73358 for a proposal of how we should fix this

LukaszMrugala replied:
I left the QEMUHandler here, if #73358 is merged first, it will be easier to incorporate.

LukaszMrugala committed:
scripts: twister: Unify Twister Statuses

Various different Statuses were joined into a single class,
TwisterStatus. This change anticipates further streamlining
of the Twister's approach to Status.

Code guarding Twister's Properties was shortened to a
value check only.

QEMUOutputStatus was left separate, as doubts were cast
whether it should remain a status. Leaving it separate makes
its removal easier.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala commented:
Comments by Anas have been incorporated, save for the QEMU - on that front we wait for https://github.com/zephyrproject-rtos/zephyr/pull/73358 to be merged.

nashif commented:
https://github.com/zephyrproject-rtos/zephyr/pull/73358 was merged last week.

LukaszMrugala commented:
> #73358 was merged last week.

Yes, I am currently rebasing this PR.

nashif commented:
@LukaszMrugala can you please rebase?

LukaszMrugala committed:
scripts: twister: Bring 3.7 changes into the new status system

Some 3.7 changes still used string statuses.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: twister: Elevate status errors

Incorrect status assignments were previously just warnings in logs.
Now they are elevated to errors.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

gchwier commented on scripts/pylib/twister/twisterlib/statuses.py:
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python3
```
nit: shebang line is rather not required here (like in many other files in twister)

gchwier reviewed (approved):
Looks good, just a comment:
do you really want to keep all the commits in the history? I know that it is good to separate some changes, but here you have changes of the same file, previous proposal etc, I found there are commits with `TestInstanceStatus`, `HarnessStatus` and so on, that were removed and are not in the final version.
